contrib_bots/lib/foursquare.py

```diff
@@ -0,0 +1,130 @@
+            @foursquare /help into the Compose Message box
+
+            Sample input:
+            @foursquare Chicago, IL
+            @foursquare help
+            '''
+
+    help_info = '''
+The FourSquare bot can receive keyword limiters that specify the location, distance (meters) and
+cusine of a resturant in that exact order.
+Please note the required use of quotes in the search location.
+
+Example Inputs:
+@foursquare 'Millenium Park' 8000 donuts
+@foursquare 'Melbourne, Australia' 40000 seafood
+                '''
+
+    def triage_message(self, message, client):
+        return '@FourSquare' in message['content']
+
+    def get_api_key(self):
+        # settings.ini must have been moved from
+        # ~/zulip/contrib_bots/lib/FourSquareBot/settings.ini into
+        # ~/settings.ini for program to work
+        # see doc.md for more information
+        with open(CONFIG_PATH) as settings:
+            config = cp.ConfigParser()
+            config.readfp(settings)
+            return config.get('FourSquare', 'api_key')
+
+    def format_json(self, text, location):
+        template = ("Food nearby {location} coming right up:")
+
+        venue_list = text['response']['venues']
+        venue_details = {}
+        for i in range(len(text['response']['venues'])):
+            template += '\n>{s}name_{i}\n>{s}formattedAddress_{i}\n>{s}descrip_{i}\n'.format(i=str(i) + '}', s='{')
+            place = venue_list[i]
+            venue_details['name_'+str(i)] = place['name']
+            venue_details['formattedAddress_'+str(i)] = ', '.join(place['location']['formattedAddress'])
+            venue_details['descrip_'+str(i)] = place['categories'][0]['pluralName']
+        venue_details['location'] = location
+        print(template)
+        return template.format(**venue_details)
+
+    def send_info(self, message, letter, client):
+        if message['type'] == 'private':
+            client.send_message(dict(
+                type='private',
+                to=message['sender_email'],
+                content=letter,
+            ))
+        else:
+            client.send_message(dict(
+                type='stream',
+                subject=message['subject'],
+                to=message['display_recipient'],
+                content=letter,
+                ))
+
+    def handle_message(self, message, client, state_handler):
+        words = message['content'].split()
+        if "/help" in words:
+            self.send_info(message, self.help_info, client)
+            return
+
+        try:
+            params = {'limit': '3'}
+            params['near']  = re.search('\'[A-Za-z]\w+[,]?[\s\w+]+?\'', message['content']).group(0)
+            params['radius'] = re.search('([0-9])\w+', message['content']).group(0)
+            params['query'] = re.search('\s(\w+)$', message['content']).group(0)
+            params['v'] = 20170108
+            params['oauth_token'] = self.api_key
+        except AttributeError:
```

Comment:
Does it make sense to handle partial success here?  In other words if some of the parameters can be parsed, we proceed?  If that's the case you'll want to re-structure this code.  If not, we may want to return early with an error.